gourmet/gtk/recipe.py

```diff
@@ -0,0 +1,787 @@
+    if text:
+        result = rec_gui.rd.parse_ingredient(text)
+        if 'rangeamount' in result:
+            result['amount'] = rec_gui.rd.rd._format_amount_string_from_amount(
+                (result['amount'], result['rangeamount'])
+            )
+            del result['rangeamount']
+        elif 'amount' in result:
+            result['amount'] = _convert.float_to_frac(result['amount'])
+        recipe_id = _check_if_recipe(rec_gui.rd, result['item'])
+        group_iter = ingredients.group_combobox.get_active_iter()
+        parent = ingredients.group_combobox.get_model().get_value(group_iter, 1)
+        ingredients.ingredients_tree.get_model().append(parent, [
+            result.get('amount', None),
+            result.get('unit', None),
+            result['item'],
+            result.get('optional', False),
+            None,
+            recipe_id,
+        ])
+        if parent:
+            ingredients.ingredients_tree.expand_row(
+                ingredients.ingredients_tree.get_model().get_path(parent),
+                False
+            )
+    ingredients.ingredient_entry.set_text('')
+    ingredients.ingredient_entry.grab_focus()
+
+
+def _check_if_recipe(db, name):
+    search = {'column': 'title', 'operator': 'LIKE', 'search': name}
+    recipes = db.search_recipes([search])
+    # NOTE: We should only ever get back at most 1 recipe, but since the db
+    # currently doesn't prevent entering duplicate recipes, we could get more.
+    # So we'll explicitly check that we only get one recipe back and only set
+    # a referencing recipe id if we do.
+    if 1 == len(recipes):
+        recipe = recipes[0]
+        assert recipe.id
+        return recipe.id
+    return None
+
+
+def _process_ingredient(recipe, ingredient, position, group_name = None):
+    result = {
+        'position': position,
+        'amount': ingredient[0],
+        'unit': ingredient[1],
+        'item': ingredient[2],
+        'optional': ingredient[3],
+        'inggroup': group_name,
+        'refid': ingredient[5],
+    }
+    if result['amount'] is not None:
+        result['amount'], result['rangeamount'] = _parse_range(result['amount'])
+    # TODO: Set the correct ingkey on edit?
+    # This uses the same logic as the original code, but it seems
+    # this would result in an incorrect ingkey if the name of the
+    # ingredient is changed?
+    if ingredient[4]:
+        assert 'ingkey' in ingredient[4] and ingredient[4].ingkey
+        result['ingkey'] = ingredient[4].ingkey
+        recipe['ingredients']['modified'].append((ingredient[4], result))
+    else:
+        result['ingkey'] = result['item'].split(';')[0]
+        recipe['ingredients']['added'].append(result)
+
+
+def _show_image(pixbuf, image, button_add, button_remove):
+    assert pixbuf
+    image.set_from_pixbuf(pixbuf)
+    image.show()
+    button_add.hide()
+    button_remove.show()
+
+
+def _generate_thumbnail(widget):
```

Comment:
These three helper functions below should be unit-tested, to ensure that there's no breakage.  

They are also duplicating some functionalities from `image_utils.py`. Perhaps they could be merged?

Reply:
I thought I had eliminated duplication with `image_utils`, although `image_utils` may perform the same functionality but with PIL. It appears that GDK Pixbufs may offer equivalent functionality, and since we already depend on the GObject infrastructure, using GDK directly means we could potentially eliminate intermediate conversions and drop the dependency on PIL altogether.